JBPM-5536: Lienzo Panel & implementations (growable,resizable)

romartin opened the pull request:
Hey @hasys 

This provides the "infinite" canvas stuff for Stunner. See https://issues.jboss.org/browse/JBPM-5536

I'm still working on the following stuff for considering this complete, but rather than this it should be ready for review and test. Pending:
* Adding more tests for new stuff
* Fix export to image and PDF (not exporting all the canvas size, just the visible area)
* Fixing the CM preview widget

Depends on
https://github.com/kiegroup/lienzo-core/pull/70
https://github.com/kiegroup/lienzo-tests/pull/53

Thanks in advance!

romartin commented:
Jenkins execute full downstream build

jomarko commented:
@romartin Hi, I played little bit manually with produced war, here are some comments/questions:

### Saving DMN diagram
- Add decision node to diagram
- Save diagram
- There appears this error in server
```
09:49:05,608 ERROR [org.drools.core.xml.ExtensibleXmlParser] (EJB default - 10) Node Info: id: name: 
Parser message: (null: 12, 41): cvc-complex-type.4: Attribute 'width' must appear on element 'dc:Bounds'.
09:49:05,608 ERROR [org.drools.core.xml.ExtensibleXmlParser] (EJB default - 10) Node Info: id: name: 
Parser message: (null: 12, 41): cvc-complex-type.4: Attribute 'height' must appear on element 'dc:Bounds'.
```

### Stunner Keyboard shortcuts
Shortcuts for adding elements don't work
- Create new Business Process
- Select the Start Event
- Press **T** to append task
- Nothing will happen

jomarko commented:
@romartin Hi, just would like to ask you to let me know when it is ready for re-review.

romartin commented:
Hey @jomarko @hasys yeah sorry guys, still working on it, finishing last fixes and missing tests. I'll let you know once this is ready for review, sorry for the inconveniences, and thanks for your feedback on this, fixing the issues you also mention @jomarko :+1:

romartin commented:
yeah @jomarko , still working here! :) 
I'm now focused on fixing the keyboard listeners and some other bugs, hope I'll close this stuff during today, so you can start the review... will let you know asap! Thanks!

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
Hey @hasys @jomarko 
Just rebased with master again, added more unit tests and did some bug fixes I found while testing. 
Soo it's ready for testing, thanks!! :+1:

romartin commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin, I tested changes in Chrome/Firefox/IE 11, and looks great, thank you. But I have some thoughts addition to [JBPM-7962](https://issues.jboss.org/browse/JBPM-7962).


- Canvas Borders doesn’t increase (see screen below)
- It is easy move your nodes out of the canvas
  - Probably there are should be no borders at all. Runtime works with negative coordinates, and when you open your process it can be scrolled to the start node, or zoomed out to see all data

![image](https://user-images.githubusercontent.com/1477262/48944000-4aa5ac00-ef25-11e8-90f1-1fcb1076c99f.png)

- When you use zoom canvas increases more than needed.
  - It is not possible to scroll to new redundant area
  - Zoom now unlimited (canvas can be zoomed out to much and you just doesn't see nodes, or zoomed in to much and you just see only several pixels of a node)
- I like to see large empty space when I model something. Just a question, is it possible to make infinite scroll? Or I also like to use Zoom with new infinite canvas approach, so probably it can be a good new way to go.
- Preview navigation
  - Click and move works good
  - Will be even better to add navigation by click (scroll canvas to the point clicked on Preview)
  - Another good possibility is select area you want to see. For example you select part of the process and canvas automatically scrolled/(and zoomed) to selected area.

Not sure what can be fixed in this PR vs fixed in future issues. But any way I think it will be very good to hear thougnts from @wrnkld. @wrnkld please, ping me if you need a shared Workbench instance with this PR.

Will check the code now.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/LienzoPanel.java:
```diff
@@ -0,0 +1,36 @@
+
+public interface LienzoPanel extends CanvasPanel {
+
+    public LienzoPanel show(final LienzoLayer layer);
```
Agree, but it is a convention for Lienzo-core. If we want to move Lienzo extensions to upstream, probably it is good to have it there.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/canvas/StunnerBoundsProviderFactory.java:
```diff
@@ -0,0 +1,46 @@
+
+public class StunnerBoundsProviderFactory {
+
+    public static final double ASPECT_RATIO = 300 / 150;
```
Integer division in floating-point context

romartin replied:
??

wrnkld commented:
Hello @hasys I am on PTO today but this deck might answer some of the "future feature" questions about the preview panel.

https://drive.google.com/open?id=1Jwxue5i-w5U7en9JQ2cytN7tmtaBZ1CT4BGmonGAzQI

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/canvas/PreviewLienzoPanelTest.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
```
Redundant suppression

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/canvas/ScrollableLienzoPanelTest.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
```
Redundant suppression

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/canvas/ScrollableLienzoPanelViewTest.java:
```diff
@@ -0,0 +1,82 @@
+    private ScrollableLienzoPanelView tested;
+
+    @Before
+    @SuppressWarnings("unchecked")
```
Redundant suppression

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/canvas/ScrollableLienzoPanelViewTest.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
```
Redundant suppression

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/canvas/StunnerBoundsProviderFactoryTest.java:
```diff
@@ -0,0 +1,72 @@
+@RunWith(LienzoMockitoTestRunner.class)
+public class StunnerBoundsProviderFactoryTest {
+
+    private Layer layer;
```
Field can be converted to a local variable

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/canvas/StunnerBoundsProviderFactoryTest.java:
```diff
@@ -0,0 +1,72 @@
+
+    private Layer layer;
+    private WiresLayer wiresLayer;
+    private WiresShape shape1;
```
Field can be converted to a local variable

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/canvas/StunnerBoundsProviderFactoryTest.java:
```diff
@@ -0,0 +1,72 @@
+    private Layer layer;
+    private WiresLayer wiresLayer;
+    private WiresShape shape1;
+    private WiresShape shape2;
```
Field can be converted to a local variable

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/test/java/org/kie/workbench/common/stunner/client/widgets/canvas/StunnerLienzoBoundsPanelTest.java:
```diff
@@ -0,0 +1,212 @@
+    private StunnerLienzoBoundsPanel tested;
+
+    @Before
+    @SuppressWarnings("unchecked")
```
Redundant suppression

romartin commented:
Hey @hasys 

Thanks for the review!  Lemme try to answer your points above:

> Canvas Borders doesn’t increase
The borders (scrollbars) should automatically increase or decrease on demand. So if you drag a node to the border it will increase, on the drag end. What is surprising to me is the screenshot you shared, as I don't see the horizontal scrollbar, right? How did you get to that point? I just tested again the application and cannot end up here, yet.

> It is easy move your nodes out of the canvas
Yeah I just implemented what you mention -  I just removed the limitations that we had before about canvas borders, now you can drag shapes even on negative locations, no limits. Isn't that working for you?

> Zoom now is unlimited
Yeah sorry I didn't mention this change. It was requested by @mdproctor as from feedback from the field.

> I  like to see large empty space when I model something....
Yeah good point... I was not thinking on this. Well the actual approach is that the canvas size is at least, as big as the current screen area, or as big as all the locations from the nodes inside the process. So I mean, if you have a few nodes, when creating a new process, it will not use too much space, because the process doesn't require it. I vote for adding this proposal in  [JBPM-7962](https://issues.jboss.org/browse/JBPM-7962), so we can analyze it with others and address it if we think it makes sense.

> Preview navigation...
About your updates on preview yeah I'm absolutely agree, although I would prefer considering these comments as improvements ( [JBPM-7962](https://issues.jboss.org/browse/JBPM-7962) ). This PR already contains lot of new stuff and IMO we should stop adding more things now and just trying to focus to integrate it.

So if you agree, I'll keep working on your comments on code and on the commented tests (sorry!),  also on the bugs when zooming, and let's add your feedback in [JBPM-7962](https://issues.jboss.org/browse/JBPM-7962) as for next steps. WDYT? Am I missing something?

Thanks in advance!

romartin commented:
BTW @hasys about the issue you mention, which makes your canvas not increase...  I've tested also in all browsers and I see on all something like this:

![screenshot from 2018-11-26 01-08-13](https://user-images.githubusercontent.com/4602417/48986035-adf33200-f10f-11e8-80cc-11b7df60455e.png)

I mean both scrollbars appears, and both grow/shrink on demand. Isn't the same for you?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/marshall/json/builder/BPMNGraphGenerator.java:
```diff
@@ -156,15 +153,6 @@ public void close() throws IOException {
         Graph<DefinitionSet, Node> graph = (Graph<DefinitionSet, Node>) factoryManager.newElement(UUID.uuid(),
                                                                                                   diagramDefinitionSetClass);
-        // TODO: Where are the BPMN diagram bounds in the Oryx json structure? Exist?
-        if (null == graph.getContent().getBounds()) {
```
This is removed because there are no "bounds" for the process, on the model side. I mean, since this point, on the model side, the process had the values for the diagram bounds (limits), but actually we don't need this anymore. Now the canvas size is unbounded, so the bounds for the process are just calculated at runtime when required. I tested also migrating to the old designer and it's all fine, there is no need to use that bounds, well in fact, there was a `TODO` here, so it was neither implemented. So removing...

romartin commented:
Hey  @hasys  @jomarko 
Just added a new commit that fixes the issues when zomming and also fixes another issue on the preview I recently found - once opening the preview panel it was not on the same scroll area as the main canvas, now it is (or should...:) )
Also added tests
thanks!

romartin commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin,

I tested some more scenarios:


- Test CM editor
  - When I added several Stages, I can’t add new Stage to the end of file, but can add in the middle of it
  - When amount of stages is reached maximum of stages at the canvas, I can’t add new stages any more
  - When I want to add User Tasks to stages on the end of the canvas I can’t do it, but when I am trying to add User tasks to some Stages in the middle, User tasks added to stages on the end (to wrong stage)
  - When I add maximum amount + 1 of User tasks to a stage, canvas enlarges in both directions: X and Y, but I am still not able to add new Stages to new free space on the right side of the canvas. So now I have spare space on right side, but not able to add Stages to it.
- Test processes modeled in old editor (despite this type of migration is not supported officially)
  - nodes in sub-sub-process are badly located.
- Test migration to old editor
  - Try to place nodes out of the canvas
    - It is not possible to use nodes out of the canvas after migration.
    - (work around: save and reopen process, all nodes will be located in one X=0 point), you can manually move all nodes to the canvas.
- Test that Canvas properties removed properly
  - looks good.
- Canvas export
  - Large process “squeezed” see attached File

![image](https://user-images.githubusercontent.com/1477262/49089588-2bc45400-f25c-11e8-82d2-0f8227bdddbd.png)

In general looks good now, thank you @romartin! Will check code now and move all notes to appropriate Jira issues, and it will be ready to go.

hasys commented:
Hi @romartin, sorry, I found one more issue mentioned by you above (I guess it can be fixed in future PRs):

When I use zoom, location of view zone on preview is not correct, see picture below: we can see nodes easily but on preview it select only empty space.

![image](https://user-images.githubusercontent.com/1477262/49090141-637fcb80-f25d-11e8-862e-3b22e0555547.png)

romartin commented:
Hey @hasys 

Great job, thanks for the review! I've fixed a couple of issues - the one you mention about invalid location of the preview zone, after zooming, and also fixed another issue found when exporting to SVG , it was just exporting the "visible" area, not all the process.

Sooo can you please give me the summary of issues you think still need to be fixed for this PR? I'm agree on opening new tickets ,as usually, but would like to close this asap... what am I still missing from that list? :)

Thanks!

romartin commented:
Also rebased as DMN it was not compiling with latest snapshots... now is everything ok.

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
hey @hasys @jomarko just rebased again

romartin commented:
Jenkins execute full downstream build

romartin commented:
Well as there were a few conflicts, just rebased with master again. Let's try

romartin commented:
Jenkins execute full downstream build

romartin commented:
The build was timeout.... let's try again

romartin commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin,

looks like build failed was caused by partial merge of https://github.com/kiegroup/jbpm-wb/pull/1236 and https://github.com/kiegroup/droolsjbpm-knowledge/pull/341 

Now it should be OK

Jenkins execute full downstream build

romartin commented:
Yeah @romartin @jomarko 

Finally WARs have been generated successfully :+1: 

So this is ready for testing. As commented please let's try to identify here which issues are related to this task and I'll work on those, on the other hand if you find other issues, not related, please just report them.

Thanks!

romartin commented:
Great job too @hasys , let's address these bugs on some near future... :+1: 
Thanks!!
